lib/evmone/instructions.hpp

```diff
@@ -890,6 +890,68 @@ inline code_iterator swapn(StackTop stack, ExecutionState& state, code_iterator
     return pos + 2;
 }
 
+inline evmc_status_code dataload(StackTop stack, ExecutionState& state) noexcept
+{
+    auto& index = stack.top();
+
+    if (state.data.size() < 32 || (state.data.size() - 32) < index)
+        return EVMC_INVALID_MEMORY_ACCESS;  // TODO better error
+
+    const auto begin = static_cast<size_t>(index);
+    uint8_t data[32] = {};
+    for (size_t i = 0; i < 32; ++i)
+        data[i] = state.data[begin + i];
+
+    index = intx::be::load<uint256>(data);
+    return EVMC_SUCCESS;
+}
+
+inline code_iterator dataloadi(StackTop stack, ExecutionState& state, code_iterator pos) noexcept
+{
+    const auto index = read_uint16_be(&pos[1]);
+
+    const auto begin = static_cast<size_t>(index * 32);
+    uint8_t data[32] = {};
+    for (size_t i = 0; i < 32; ++i)
+        data[i] = state.data[begin + i];
+
+    stack.push(intx::be::load<uint256>(data));
+    return pos + 3;
+}
+
+inline evmc_status_code datacopy(StackTop stack, ExecutionState& state) noexcept
+{
+    const auto& mem_index = stack.pop();
+    const auto& data_index = stack.pop();
+    const auto& size = stack.pop();
+
+    if (!check_memory(state, mem_index, size))
+        return EVMC_OUT_OF_GAS;
+
+    auto s = static_cast<size_t>(size);
+
+    if (state.data.size() < s || state.data.size() - s < data_index)
+        return EVMC_INVALID_MEMORY_ACCESS;  // TODO better error
+
+    const auto copy_cost = num_words(s) * 3;
```

Comment:
Hm, we could have some helper for this, given it is used by so many instructions.

Reply:
@gumb0 how about this?

Reply:
By helper I mean `copy_cost` calculation / check.

Reply:
Added a helper.